### PR TITLE
Github #31984: fix for array offset bug during some 404s (PHP 7.4)

### LINKED ISSUE #31984: Expecting 404, getting 500: "Trying to access array offset on value of type bool"
### Preconditions (*)

1. Magento 2.4.1
2. PHP 7.4

### Steps to reproduce (*)

1. As an admin, create an saleable product with any URL, such as "box"
2. Set "Generate "category/product" URL Rewrites" to "No"
3. As a customer, visit the URL /abc/box.html

### Expected result (*)

1. 404 page not found error, as the category abc does not exist

### Actual result (*)

1. 500, Trying to access array offset on value of type bool on vendor/magento/module-catalog-url-rewrite/Model/Storage/DynamicStorage.php on line 182

PHP 7.4 is become stricter than PHP 7.3 with regards to these types of errors and that's why this bug is showing up it's head now.

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php#L182

The way the code is written doesn't make sense: the false check is done after an array access is attempted; the false check should be done first.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by erfanimani**
### Description (*)

Fixes "Trying to access array offset on value of type bool" when accessing certain non existing category paths for known products, on PHP 7.4.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#31984

### Manual testing scenarios (*)

1. On PHP 7.4, as an admin, set "Generate "category/product" URL Rewrites" to "No"
2. Make a GET request to {nonexistentcategory}/{product_url} (where product_url is an existing saleable product)
3. Ensure status code is 404, not 500

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @erfanimani,
Could you cover your changes with any type of automated tests?

**Comment by erfanimani** [comment]
@ihor-sviziev I'm thinking it might be hard to isolate this particular issue as it's nested within a few private method calls, but I'll have a look tomorrow — thanks for the feedback

**Comment by erfanimani** [comment]
Hi @ihor-sviziev, I can't create unit tests as the SUT contains only private/protected methods. I can create a MFTF test that reproduces the error (although sometimes it won't, depending on PHP error reporting and PHP version), and simply returns a 404 not found if successful. I'm not sure how relevant a test like that would be for the future. What's your opinion?

See the original issue for steps to reproduce.

**Comment by ihor-sviziev** [comment]
I don’t think mftf test sis suitable here. 
@engcom-Charlie do you feel we can cover this change with some unit / integration test? If so - could you share your thoughts how to do that?

**Comment by engcom-Charlie** [comment]
@ihor-sviziev I think can try via unit test by using reflection.
Smth like this https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Paypal/Test/Unit/Model/Payflow/Service/GatewayTest.php#L206

**Comment by ihor-sviziev** [comment]
TBH I have a feeling that this logic should be extracted to a separate model and tested separately. What do you think?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @erfanimani,
Could you cover your changes with the unit test, as suggested in https://github.com/magento/magento2/pull/31985#issuecomment-778054763 ?

**Comment by erfanimani** [comment]
Thanks @ihor-sviziev, I'll try and update the PR.

**Comment by erfanimani** [comment]
@ihor-sviziev, please see the updated PR. I believe I've got 100% test coverage of the method that was changed. Feedback is appreciated.

Thanks.

**Comment by erfanimani** [comment]
Oops, closed the PR by accident. Reopened.

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests, Static Tests

File: app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+{
+    /**
+     * @var DynamicStorage
+     */
+    private $object;
+
+    /**
+     * @var UrlRewriteFactory|MockObject
+     */
+    private $urlRewriteFactory;
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php]: Please add "Mock" suffix to all mocks
```suggestion
    private $urlRewriteFactoryMock;
```

File: app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+     * @dataProvider findProductRewriteByRequestPathDataProvider
+     */
+    public function testFindProductRewriteByRequestPath(
+        $data,
+        $productFromDb,
+        $categorySuffix,
+        $categoryFromDb,
+        $canBeShownInCategory,
+        $expectedProductRewrite
+    ) {
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php]: Let's define object types 
```suggestion
    public function testFindProductRewriteByRequestPath(
        array $data,
        $productFromDb,
        string? $categorySuffix,
        $categoryFromDb,
        bool $canBeShownInCategory,
        array? $expectedProductRewrite
    ): void {
```

File: app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        );
+    }
+
+    /**
+     * @param $data
+     * @param $productFromDb
+     * @param $categorySuffix
+     * @param $categoryFromDb
+     * @param $canBeShownInCategory
+     * @param $expectedProductRewrite
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php]: please define expected types for params

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by erfanimani** [comment]
Thanks for the feedback @ihor-sviziev, I've updated the PR with your suggestions.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run  Integration Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8884](https://jira.corp.magento.com/browse/ENGCOM-8884) has been created to process this Pull Request

**Comment by dmitriyprime** [comment]
:heavy_check_mark:  **QA passes**

:x: **Before:**
before

:heavy_check_mark: **After**
after

**Pull request merged**